# Library Import

* Importing the necessary libraries for my project

In [130]:
#import pandas
import pandas as pd
#import Numpy
import numpy as np

# Loading the data into a Dataframe

* Using the pandas function 'read_csv' to read the data from the file location

In [131]:
#Loading the dataset into the heart_disease library
heart_disease = pd.read_csv ("/Users/mac/Downloads/011 heart-disease.csv")

## Checking the first few records in the new Dataframe

* The head() function will be used to check the first few record in the newly loaded dataframe

In [132]:
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

## Choosing Features and Target variables

* Features is achieved by removing only the target column from the dataframe and assigning it to variable X
* Target is achieved by just assigning the data in the target column to the variable Y

In [133]:
# Features as X
X = heart_disease.drop("target",axis = 1) 

# Target as Y
y = heart_disease["target"]

## Choosing the model and hyperparameter

* Random Forest will be used for this task

In [134]:
#importing the Randomforest library
from sklearn.ensemble import RandomForestClassifier

#initiating the model
clf = RandomForestClassifier(n_estimators = 100)

#getting the hyperparameters
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

## Data Splitting

* Split the model into training and test data

In [135]:
#import train test split 
from sklearn.model_selection import train_test_split

#Splitting the data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

## Fitting the model
* The fit function from Randomclassifiers will be used

In [136]:
# Fitting the model
clf.fit(X_train, y_train);

## Make a prediction
 
* Make the prediction using X_test and using the predict function from RandomClassifier


In [137]:
#predict using the 
y_predicts = clf.predict(X_test)
y_predicts

array([0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0])

* Result of the prediction is in the form of an array but it corresponds to our target variable which has 1's and 0's

## Evaluate the model
* The model will be evaluated using the score function from Randomforestclassifier
* And also using classification report, confusion matrix and accuracy score

In [138]:
# Evaluating the score of the training data
clf.score(X_train, y_train)

1.0

In [139]:
# Evaluating the score of the test data
clf.score(X_test, y_test)

0.8852459016393442

* The training gives a 1.0(100%) score because it trained on a seen data. While the test data give a score of 0.885(89%) which means it can predict the target at a 89% accuracy

In [140]:
# importing libraries for classification report, confusion matrix and accuracy score
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [141]:
#printing the result of classification report of the target test and the predicted test
print(classification_report(y_test, y_predicts))

              precision    recall  f1-score   support

           0       0.90      0.87      0.88        30
           1       0.88      0.90      0.89        31

    accuracy                           0.89        61
   macro avg       0.89      0.88      0.89        61
weighted avg       0.89      0.89      0.89        61



In [142]:
#printing the result of confusion matrix report of the target test and the predicted test
print(confusion_matrix(y_test, y_predicts))

[[26  4]
 [ 3 28]]


In [143]:
#printing the result of accuracy score report of the target test and the predicted test
print(accuracy_score(y_test, y_predicts))

0.8852459016393442


## Improving the model

* Since the Accuracy of the model is 89%, there's room for improvement
* A good way of doing this is to use different n_estimator on the algorithm

In [144]:
np.random.seed(42)
for i in range(10, 100, 10):
    print(f"Trying model with {i} Estimator")
    #training the model
    clf = RandomForestClassifier(n_estimators = i).fit(X_train,y_train)
    #printing the accuracy for each value of estimator
    print(f"Model Accuracy on test set: {clf.score(X_test, y_test) * 100:.2f} %")
    print(" ")

Trying model with 10 Estimator
Model Accuracy on test set: 78.69 %
 
Trying model with 20 Estimator
Model Accuracy on test set: 85.25 %
 
Trying model with 30 Estimator
Model Accuracy on test set: 81.97 %
 
Trying model with 40 Estimator
Model Accuracy on test set: 86.89 %
 
Trying model with 50 Estimator
Model Accuracy on test set: 83.61 %
 
Trying model with 60 Estimator
Model Accuracy on test set: 83.61 %
 
Trying model with 70 Estimator
Model Accuracy on test set: 81.97 %
 
Trying model with 80 Estimator
Model Accuracy on test set: 86.89 %
 
Trying model with 90 Estimator
Model Accuracy on test set: 80.33 %
 


### From the various estimators, none of the estimator was able to improve our 89% accuracy. Which means 100 n_estimator is good for the Random algorithm

## Save and load the model

* the model can be saved using pickle.dump function
* the model can be opened using pickle.load function

In [145]:
#importing the library
import pickle

#writing the model into a file
pickle.dump(clf, open("/Users/mac/Downloads/random_forest_model_1.pk1", "wb"))

In [146]:
#Reading model from a file

loaded_model = pickle.load(open("/Users/mac/Downloads/random_forest_model_1.pk1", "rb"))

#Testing the loaded model
loaded_model.score(X_test , y_test)

0.8032786885245902